Astropy CSV table reader using pyarrow

taldcroft opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is an implementation of a fast CSV reader for astropy that uses [pyarrow.csv.read_csv](https://arrow.apache.org/docs/python/csv.html). This was discussed in #16869. In particular, a speed-up of around 10x over pandas and the astropy fast reader is noted in this [profiling](https://github.com/astropy/astropy/issues/16869#issuecomment-2307882059) by @dhomeier.

The goal for this reader is to make an interface that will be familiar to astropy `io.ascii` users, while exposing some additional features brought by pyarrow `read_csv`. The idea is to keep the interface clean and consistent with astropy.

A quick demonstration notebook that you can use to play with this is at: https://gist.github.com/taldcroft/ac15bc516a7bf7c76f9eec644c787298

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #16869

### Related

https://github.com/pandas-dev/pandas/issues/54466

<!-- Optional opt-out -->

Please DO squash and merge, the individual commits are not valuable here.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

pllim commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    encoding: str = "utf-8",
+    newlines_in_values: bool = False,
+) -> "Table":
+    """Read a CSV file into an astropy Table using pyarrow.csv.read_csv().
```
```suggestion
    """Read a CSV file into an astropy Table using ``pyarrow.csv.read_csv()``.
```

pllim commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    ----------
+    input_file : str, PathLike, or binary file-like object
+        File path or binary file-like object to read from.
+    delimiter: 1-character str, optional (default ",")
```
Same suggestion to all similar occurrences. 

```suggestion
    delimiter : 1-character str, optional (default ",")
```

pllim reviewed (commented):
Thanks! I want to benchmark this but does that mean we need to install pyarrow in https://github.com/astropy/astropy/blob/main/.github/workflows/ci_benchmark.yml ?

taldcroft commented:
> Thanks! I want to benchmark this but does that mean we need to install pyarrow in https://github.com/astropy/astropy/blob/main/.github/workflows/ci_benchmark.yml ?

There are one-time benchmarks here: https://github.com/astropy/astropy/issues/16869#issuecomment-2307882059. These demonstrate that pyarrow `read_csv()` appears to be a factor of 10 faster than any other readers.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    """
+    pa, _ = get_pyarrow_csv()
+
+    if pa.types.is_string(arr.type):
```
Not necessarily better, but you could consider doing here:
```
if pa.types.is_string(arr.type):
    def to_numpy(arr):
        return arr.to_numpy().astype(str)
else:
    def to_numpy(arr);
        return arr.to_numpy()
```
and then the rest of the routine no longer has to worry about str vs numerical.

Or, maybe easier, combining with what I suggest below,
```
if arr.null_count == 0:
   # No, nulls, just get an ndarray view of the pyarray
    np_array = arr.to_numpy()
else:
   # Create a Masked Array copy, with nulls filled with zero.
    np_array = np.ma.array(arr.fill_null(0).to_numpy(), mask=arr.is_null().to_numpy(), copy=False)

return np_array.astype(str) if pa.types.is_string(arr.type) else np_array
```

mhvk replied:
Aside: `arr.type == "string"` seems to work - means no import is necessary.

taldcroft replied:
Done.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    return pa, csv
+
+
+def convert_pa_string_array_to_numpy(str_arr: "pa.ChunkedArray") -> np.ndarray:
```
Don't think the separate routine is really necessary, see in the main one.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+
+def get_pyarrow_csv():
+    """Helper function to import pyarrow and pyarrow.csv."""
+    from astropy.utils.compat.optional_deps import HAS_PYARROW
```
I think this can be on top - that does not actually import but just checks whether something is present.

I'd probably move the function up too.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    """
+    pa, csv = get_pyarrow_csv()
+
+    parse_options = pa.csv.ParseOptions(
```
Might as well use `csv` here. That may mean `pa` itself is not needed here.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    """
+    from astropy.table import Table
+
+    columns = {}
```
Can be written as an dict comprehension, maybe even slightly clearer:
```
return Table(
    {
        "name": convert_pa_array_to_numpy(col)
        for name, col in zip(table_pa.column_names, table_pa.itercolumns())
    },
    copy=False,
)
```

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    return out
+
+
+def strip_comment_lines(
```
I guess a question will be whether this relatively slow copy is actually worth it, or whether it is better to just fall back to the default reader for this case. After all, the most common case, of comments at the start of the file, can already be dealt with. My sense would be not to do this in the initial PR.

Note in this respect that for things of interest to reading with `pyarrow`, it is not impossible that the file size is of order the amount of memory one has!

taldcroft replied:
For the moment let's leave this in. One idea I've had is writing a streaming FileLike class that strips comments on a line-by-line basis. My first attempt at this failed within pyarrow somehow and I didn't track it down, but this *should* be possible. Then you might take a hit on speed but not memory.

Also don't forget a potential use case of reading thousands of big (but not huge) CSV files. E.g. reading the raw GAIA archive which they make available in ECSV... though of course I'd like to write a `pyarrow.ecsv` reader!

taldcroft replied:
I did some simple profiling, creating a CSV file that is 100000 lines with a few comments at the top. Even with this current implementation, the `pyarrow.csv` reader is about 35% faster than the fast ASCII reader. 

There might also be ways to optimize comment handling, for instance doing a first pass looking for commented lines without keeping the entire file in memory. If the commented lines are all at the top (presumably the most common case) then the pyarrow reader can be set to just skip those lines.

mhvk reviewed (commented):
Nice! I like the general idea; my only more major comment is that I'm not sure in this initial stage one should add the commented-line skipper.

For follow-up, I guess, would be to make this the default "first try" if `pyarrow` is available, and then deprecate the fast reader?

It does seem `Table.{from,to}_pyarrow` methods would be reasonable, but better as follow-up.

hamogu commented:
Since @dhomeier has shown pyarrow to be significantly faster, it would be good to have it for the biggest tables. And this is a relatively thin wrapper just to match the API we are used to, so why not?
I do wonder (similar to @mhvk ) how far it makes sense to go to have capabilities that are not native to pyarrow (e.g. comment characters). Is it worth the pure-Python preprocessing at all? Would that dilute the advertised point "this is super fast and super-memory efficient, so use it for tables in the GB range"?

For smaller tables we have other established solutions which are more flexible (not the least our own pure-python readers and our own C reader). How many GB-sized tables are there in the wild with commented lines that are not in the header? I'm just worried about user confusion along the lines of "It's reading this table just fine and that table that's almost identical (but with comment lines) crashes with a Python out-of-memory error". Of course, that only applies to the biggest tables of them all. For csv files in the 0.5-1GB rage, this is probably still be faster AND would fit into memory (and maybe not be too slow) on modern machines. So it's a trade-off.

taldcroft commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    return pa, csv
+
+
+def convert_pa_string_array_to_numpy(str_arr: "pa.ChunkedArray") -> np.ndarray:
```
Done, but in this case the output type is not known.

taldcroft commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,377 @@
+    if pa.types.is_string(arr.type):
+        return convert_pa_string_array_to_numpy(arr)
+
+    # Check if array has any nulls
```
Done.

taldcroft reviewed (commented):
Thanks for the great comments! I think I've addressed them all, or at least responded. Sounds like I have agreement to keep going ahead on this and start working on tests, docs etc?

neutrinoceros commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,356 @@
+
+
+def get_convert_options(
+    include_names: list | None,
```
```suggestion
    include_names: list[str] | None,
```

neutrinoceros commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,356 @@
+def get_convert_options(
+    include_names: list | None,
+    dtypes: dict[str, "npt.DTypeLike"] | None,
+    null_values: list | None,
```
```suggestion
    null_values: list[str] | None,
```

neutrinoceros commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,356 @@
+    include_names: list[str] | None = None,
+    dtypes: dict[str, "npt.DTypeLike"] | None = None,
+    comment: str | None = None,
+    null_values: list | None = None,
```
```suggestion
    null_values: list[str] | None = None,
```

taldcroft reviewed (commented):
@mhvk @neutrinoceros - I've addressed all the first round of comments. 

I lost track of the feature freeze date until the recent announcement, but I hope this can get into this release. I'm planning to add testing and documentation on Monday so I would be very grateful if you are able to make time to look at this again early this week.

taldcroft commented:
One unfortunate thing I just noticed is that my current profiling shows that this full implementation is only a factor of 2x faster than the fast ASCII reader. From the original profiling It looks suspiciously like multi-threading is not being enabled, but as far as I can tell that is not the case.
```
dat = simple_table(size=100000, cols=20)
dat.write("junk.csv", format="ascii.basic", overwrite=True, delimiter=",")

%timeit Table.read("junk.csv", format="ascii.csv", fast_reader="force", delimiter=",", guess=False)
103 ms

%timeit Table.read("junk.csv", format="pyarrow.csv")
51 ms
```
I'm not sure if there is some other significant overhead being introduced here, but I'll need to do some profiling.

mhvk commented:
> It turns out that converting an object arrays of strings into a numpy string array is the bottleneck.

This may well go away eventually, with numpy's new `StringDType` - there's talk about trying to have a copy-less version with pyarrow. Anyway, a factor 2 is still nice! And presumably a lot better for mostly numeric data?

taldcroft commented:
Profiling the new implementation `pyarrow.csv` vs fast `io.ascii` gives:
- About 30% faster for all string columns
- About 16x faster for all int and float columns

For a table with two commented lines at the top of the file:
- About 5x faster for all int, float columns, with no change in memory use.

taldcroft commented:
For a real-world example, reading the first Gaia ECSV source file here: https://cdn.gea.esac.esa.int/Gaia/gdr3/gaia_source/ (after unzipping) gives:
- fast `io.ascii`: 10.1 s (using the CSV reader just ignoring comments)
- `pyarrow.csv`: 1.48 s

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+    encoding: str = "utf-8",
+    newlines_in_values: bool = False,
+) -> "Table":
+    """Read a CSV file into an astropy Table using ``pyarrow.csv.read_csv()``.
```
Total nitpick, but I think for docstrings it is best to stick to wrapping at <80 characters; these are seen on terminals (and, heck, it is just annoying to see the breaks on the github interface...)

taldcroft replied:
I'm not sure where you are seeing the line breaks. All the docstrings here are wrapped at 88 characters following the black standard. When I look at this on GitHub it seems fine. Here is an example docstring that has exactly 88 characters in one line:
<img width="707" alt="image" src="https://github.com/user-attachments/assets/2b038e1d-11e4-4a4d-9f01-0e4e392861dc" />

mhvk replied:
Must be that I'm using a vertical monitor then...
![image](https://github.com/user-attachments/assets/915b238a-aadb-4c69-bb46-74fc0a99bf7d)

Feel free to ignore!

taldcroft replied:
> Must be that I'm using a vertical monitor then...

I can replicate that by making my browser narrow, but at that width then some of code also gets wrapped badly. So I think it's just you and I'll ignore. :smile:

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+        read_options=read_options,
+        convert_options=convert_options,
+    )
+    table_apt = convert_pa_table_to_astropy_table(table_pa)
```
Nitpick: just return it directly.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+        PyArrow CSV ConvertOptions object configured with the specified column names and
+        data types.
+    """
+    pa, csv = get_pyarrow_csv()
```
This is the only place `pa` is used as well as `csv`. I would suggest to have `get_pyarray_csv` only return `csv`, and here add `from pyarrow import from_numpy_dtype` and use that below.

taldcroft replied:
This function is very much the same pattern as https://github.com/taldcroft/astropy/blob/061161091f861a64010f6ac9087a2fec0daa03a9/astropy/io/misc/parquet.py#L507.

My worry about changing this now is that there may be code in the future that only needs `pa`, and then you need to revert this or duplicate the import exception handling in a new function. For instance I definitely want to write a `pyarrow.ecsv` reader, and it is certainly possible that will have a function that requires a `pyarrow` import without `csv`.

taldcroft replied:
Actually, I have a different idea. Stand by... :smile:

taldcroft replied:
Fixed, hopefully.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+    return table_apt
+
+
+def get_pyarrow_csv():
```
See my suggestion below to let this return `csv` only. This would be  as the name suggests (though if could be `from_pyarrow_import_csv` or `import_pyarrow_csv_or_raise` for greater clarity).

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+        else:
+            out = arr.to_numpy()
+    else:
+        # Return a Masked Array copy, with nulls filled with zero.
```
Do we care what is underneath masked values? For numerical data at least, I think not, which means we can just have a masked array view.
Alternatively, can we just *set* the values to zero? It is not as if the pa array somehow needs to be preserved. Or are pa arrays read-only?

taldcroft replied:
I'm definitely not an expert on pyarrow, but it does appear that the array is read-only:
```
In [1]: import pyarrow as pa
   ...: arr = pa.array([1, 2, None, 3], type=pa.int8())
   ...: fill_value = pa.scalar(5, type=pa.int8())

In [2]: arr
Out[2]: 
<pyarrow.lib.Int8Array object at 0x10e21b8e0>
[
  1,
  2,
  null,
  3
]

In [3]: arr[arr.is_null()] = fill_value
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 arr[arr.is_null()] = fill_value

TypeError: 'pyarrow.lib.Int8Array' object does not support item assignment
```

About the "masked array view" comment, I'm not sure what you meant by that.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+
+    Notes
+    -----
+    - If the input array is of string type, it delegates the conversion to
```
This comment is out of date and can be removed.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+        name: convert_pa_array_to_numpy(col)
+        for name, col in zip(table_pa.column_names, table_pa.itercolumns())
+    }
+    out = Table(columns, copy=False)
```
I'd again tend to just return directly here, but obviously no big deal.

taldcroft replied:
I use this pattern a lot because it facilitates debugging. From the context of this function I can easily examine `out` along with the other variables defined in the function.

mhvk replied:
OK, that is a nice reason. We better make sure we never enable the ruff rule that would remove it!

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,402 @@
+        # If we get here, we need to read the whole file and remove comment lines and
+        # write to an output BytesIO file.
+        output = io.BytesIO()
+        for idx, line in enumerate(input_file):
```
Since we might have been reading some lines above, don't we need to something like `input_file.seek(0)`? Be sure to add an appropriate test case!

taldcroft replied:
I agree and good catch, I'll definitely test that.

mhvk reviewed (commented):
@taldcroft - this mostly looks good, and quite a few of my comments are really just nitpicks. Though I think I found a bug that more detailed tests would surely have found too...

p.s. The GAIA example shows this is really worth it, great!

taldcroft commented:
@mhvk - I added a suite of tests that covers all the `read_csv` arguments and a few more things. 

A significant addition since you last looked is support for date, time, timestamp types. Currently these result in numpy datetime64 arrays, or an object array of datetime.time objects (for pure times like 12:34:45.233). An obvious idea is an option to allow converting to a `Time` mixin column where applicable.

Apart from that, I think this is mostly feature complete at least for this release. So it's just the docs, change log, what's new.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,482 @@
+    elif pa.types.is_timestamp(arr.type):
+        fill_value = pa.scalar(datetime.datetime(2000, 1, 1), type=arr.type)
+    else:
+        raise TypeError(f"unsupported PyArrow array type: {arr.type}")
```
Are there any unsupported `PyArrow` types left?  If not, the above logic is just for `fill_value`, which is not needed if there are no masked elements, so it could be moved inside the masked branch.

mhvk commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -0,0 +1,482 @@
+        # No nulls, just return an ndarray view of the pyarray
+        out = convert_pa_string_array_to_numpy(arr) if is_string else arr.to_numpy()
+    else:
+        # Fill nulls in `arr` with zero. We do not know of a zero-copy fill for
```
I still slightly wonder why we care about filling the null elements. Is it not good enough that they are masked? (except perhaps for string, where one wouldn't want one very long masked string drive the output array size; but that needs a copy anyway!)

Alternatively, if you think setting a fill value is important, then from a quick check it looks like the array made with `to_numpy()` is writeable, so maybe an idea is just to fill the numpy array? I.e., something like,
```
mask = arr.is_null().to_numpy(zero_copy_only=False)
if is_string:
    data = convert_pa_string_array_to_numpy(arr.fill_null(fill_value))
else:
    data = arr.to_numpy()
    data[mask] = fill_value
out = np.ma.array(data, mask=mask, copy=False)
```

Quick speed test:
```
import numpy as np
from astropy.table import Table
from pyarrow import csv

ma = np.ma.arange(1000.0)
ma[500] = np.ma.masked
Table([ma], names=["ma"]).write("a.csv", overwrite=True)

p = csv.read_csv("a.csv")["ma"]
p
# <pyarrow.lib.ChunkedArray object at 0x7f6fb8051180> ...

def to_numpy(p, fill_value=0.):
    m = p.is_null().to_numpy(zero_copy_only=False)
    a = p.fill_null(fill_value).to_numpy()
    return np.ma.MaskedArray(a, mask=m, copy=False)

def to_numpy2(p, fill_value=0.):
    m = p.is_null().to_numpy(zero_copy_only=False)
    a = p.to_numpy()
    a[m] = fill_value
    return np.ma.MaskedArray(a, mask=m, copy=False)

%timeit to_numpy(p)
72.6 μs ± 97.8 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%timeit to_numpy2(p)
46.4 μs ± 123 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

So, it seems worth it!

Though on second thought, perhaps not so much for speed - reading itself takes far longer - but rather to avoid extra memory usage.

taldcroft replied:
I should have put a code comment why pre-filling the `pa` array is required. Basically conversion does not work without it, at least for what I have tried. In this case it casts the Int array to float and uses `NaN` for missing data.
```
In [7]: x = pa.array([1,None, 2])

In [8]: x.to_numpy(zero_copy_only=False)
Out[8]: array([ 1., nan,  2.])
```

mhvk replied:
Ah, I see, so it was a bit of bad luck that I happened to try a float array where this is not an issue. Let's just keep what you have, then; if there is a pressing need, one could always still special-case `float`.

mhvk commented on astropy/io/misc/pyarrow/tests/test_csv.py:
```diff
@@ -0,0 +1,605 @@
+    for col1, col2 in zip(t1.itercols(), t2.itercols()):
+        assert col1.name == col2.name
+        assert col1.dtype == col2.dtype
+        assert np.array_equal(col1, col2)
```
I'd use `numpy.testing`'s `assert_array_equal` - gives nicer error message on failure, pointing out where the discrepancy is.

taldcroft replied:
Done

mhvk reviewed (commented):
This looks rather nice! One real comment left, with the suggestion to do the filling on the numpy side for speed and to save memory. But even that not serious.

A problem though is that most CI runs errored:
```
TypeError: ChunkedArray.to_numpy() takes no keyword arguments
```

taldcroft commented:
@mhvk - I think this ready for final review. 

- Fixed the annoying Windows-only problems.
- Fixed a real bug in the comment handling and added a number of tests.
- Fixed most of the coverage issues. The remaining 3 bits of uncovered code are OK (import check and two data types where the code is obvious).
- Added a lot more documentation.

taldcroft commented:
@mhvk - gentle reminder, and with apologies for being late on this, today is the feature freeze deadline. :smile:

mhvk commented on docs/io/ascii/index.rst:
```diff
@@ -34,9 +38,14 @@ section on `Supported formats`_ contains the full list.
+.. note::
+
+   For reading large CSV files, the Table :ref:`PyArrow <table_io_pyarrow_csv>` interface
+   is an option to consider since it can be up to 15 times faster than other readers. For
```
Just "is recommended"? If not that, either change the wording here or in the second line, to avoid repeating "an option to consider"

taldcroft replied:
This second note was actually orphaned and I had intended to cut it. It's done now, with the Pandas-for-writing moved up as well.

taldcroft replied:
New note (with the 2nd one removed):
<img width="762" alt="image" src="https://github.com/user-attachments/assets/eacf411d-83a6-4db0-9c2a-73f698931f35" />

mhvk commented on docs/io/misc.rst:
```diff
@@ -1,5 +1,5 @@
 **************************************************************
-Miscellaneous: HDF5, YAML, Parquet, pickle (`astropy.io.misc`)
+HDF5, Parquet, pickle, PyArrow CSV, YAML (`astropy.io.misc`)
```
I just checked and nowhere in the links in this document is `pickle` used - I'd suggest to just remove it here while you are at it.

taldcroft replied:
OK, I see what happened. There are some pickle functions but they have been deprecated so the API docs got taken away. Good catch, I have taken `pickle` out of the title.

mhvk commented on docs/io/unified_table_text.rst:
```diff
@@ -16,6 +16,12 @@ powerful interface for reading and writing text tables. In addition, the interfa
+.. note::
+
+   For reading large CSV files, the astropy :ref:`PyArrow CSV <table_io_pyarrow_csv>`
+   reader is a good option to consider since it can be up to 15 times faster than other
```
Seeing it again, I think "is recommended" is the way to go here!

taldcroft replied:
On this matter I think there are many valid considerations that might influence the decision each user makes. Maybe they don't want to install pyarrow, or maybe `io.ascii` is already working well enough, or maybe they use some feature of the pandas reader. So here the goal is to call attention to this option without making a ranked value judgement.

mhvk reviewed (approved):
@taldcroft - I have had another look, now focussing on the documentation. My remaining comments are tiny and optional, so I'll approve now. A very nice addition!

p.s.  Longer term, I guess a question is whether to deprecate the fast reader - my tendency would be to do so, but obviously that's really for later.

taldcroft commented:
@mhvk - thanks so much for the review, as always this PR ended up in a much better place! I've addressed your minor docs comments and auto-merge squash is now enabled. 🤞 it gets in without another What's New merge conflict.
